File: typescript/src/client/RefusalApi.ts
/**
 * Parmana Refusal Record API (RFC-0021).
 */

import type {
  RefusalRecord,
} from "../models/index.js";

import type {
  Transport,
} from "../config/Transport.js";

export class RefusalApi {
  constructor(
    private readonly transport: Transport,
  ) {}

  /**
   * Verifies a Refusal Record's signature. Maps to POST
   * /refusal/verify -- takes the record itself, not a lookup by id.
   * No caller authentication is required by this route: it is the
   * capability that makes a refusal independently third-party
   * verifiable (RFC-0021).
   */
  public async verify(
    record: RefusalRecord,
  ): Promise<boolean> {
    const response =
      await this.transport.send<{ valid: boolean }>({
        method: "POST",
        path: "/refusal/verify",
        body: record,
      });

    return response.body.valid;
  }

  /**
   * Retrieves a Refusal Record by Business Transaction ID. Maps to
   * GET /refusal/:businessTransactionId. Unlike verify() above, this
   * route is behind caller-auth and ownership scoping, identically to
   * TrustRecordApi.get().
   */
  public async get(
    businessTransactionId: string,
  ): Promise<RefusalRecord> {
    const response =
      await this.transport.send<RefusalRecord>({
        method: "GET",
        path: `/refusal/${businessTransactionId}`,
      });

    return response.body;
  }
}

File: python/parmana/api/refusal_api.py
"""
Parmana Refusal Record API (RFC-0021).
"""

from __future__ import annotations

from typing import cast

from parmana.config.transport import Transport
from parmana.models.refusal_record import RefusalRecord
from parmana.serialization import encode


class RefusalApi:
    """
    Refusal Record API (RFC-0021).

    Responsibilities
    ----------------
    - Verify a Refusal Record's signature
    - Retrieve a Refusal Record by Business Transaction ID

    This API does NOT:
    - execute Business Transactions
    - verify Execution Trust Records
    - replay executions
    """

    def __init__(
        self,
        transport: Transport,
    ) -> None:
        self._transport = transport

    def verify(
        self,
        record: RefusalRecord,
    ) -> bool:
        """
        Verify a Refusal Record's signature.

        Maps to POST /refusal/verify -- takes the record itself, not a
        lookup by id. No caller authentication is required by this
        route: it is the capability that makes a refusal independently
        third-party verifiable (RFC-0021).

        Parameters
        ----------
        record:
            The Refusal Record to verify.

        Returns
        -------
        True if the signature is valid, False otherwise.
        """

        payload = cast(
            "dict[str, bool]",
            self._transport.send(
                method="POST",
                path="/refusal/verify",
                body=encode(record),
            ),
        )

        return payload["valid"]

    def get(
        self,
        business_transaction_id: str,
    ) -> RefusalRecord:
        """
        Retrieve a Refusal Record by Business Transaction ID.

        Maps to GET /refusal/:businessTransactionId. Unlike verify()
        above, this route is behind caller-auth and ownership scoping,
        identically to TrustRecordApi.get().

        Parameters
        ----------
        business_transaction_id:
            Business Transaction identifier.

        Returns
        -------
        Refusal Record.
        """

        return self._transport.send(
            method="GET",
            path=f"/refusal/{business_transaction_id}",
            response_model=RefusalRecord,
        )
